src/org/ods/component/ScanWithSnykStage.groovy

```diff
@@ -0,0 +1,88 @@
+package org.ods.component
+
+import org.ods.services.SnykService
+
+class ScanWithSnykStage extends Stage {
+  protected String STAGE_NAME = 'Snyk Security Scan'
+  protected SnykService snyk
+
+  ScanWithSnykStage(def script, IContext context, Map config, SnykService snyk) {
+    super(script, context, config)
+    if (!config.containsKey('failOnVulnerabilities')) {
+      config.failOnVulnerabilities = context.failOnSnykScanVulnerabilities
+    }
+    if (!config.organisation) {
+      config.organisation = context.projectId
+    }
+    if (!config.projectName) {
+      config.projectName = context.componentId
+    }
+    if (!config.buildFile) {
+      config.buildFile = 'build.gradle'
+    }
+    this.snyk = snyk
+  }
+
+  def run() {
+    if (!config.snykAuthenticationCode) {
+      script.error "Option 'snykAuthenticationCode' is not set!"
+    }
+
+    def message = "Snyk scan mode: build will " + (config.failOnVulnerabilities ? "" : "not ") +
+      "fail if vulnerabilities are found (failOnVulnerabilities=${config.failOnVulnerabilities})!"
+    script.echo message
+
+    if (!snyk.version()) {
+      script.error 'Snyk binary is not in $PATH'
+    }
+
+    if (!snyk.auth(config.snykAuthenticationCode)) {
+      script.error 'Snyk auth failed'
+    }
+
+    boolean testSuccess
+
+    def envVariables = [
+      "NEXUS_HOST=${context.nexusHost}",
+      "NEXUS_USERNAME=${context.nexusUsername}",
+      "NEXUS_PASSWORD=${context.nexusPassword}"
+    ]
+    script.withEnv(envVariables) {
+      if (!snyk.monitor(config.organisation, config.buildFile, config.projectName)) {
+        script.error 'Snyk monitor failed'
+      }
+
+      testSuccess = snyk.test(config.organisation, config.buildFile)
+      if (testSuccess) {
+        script.echo 'No vulnerabilities detected.'
+      } else {
+        script.echo 'Snyk test detected vulnerabilities.'
+      }
+    }
+
+    generateAndArchiveReport()
```

Comment:
I think we need to add here a `try / catch` block to capture any error and log the details.